### PR TITLE
!!! FEATURE: UI rewrite to Fusion/React & new fields & export/import

### LINKED ISSUE #3: Import & export file with redirect entries
It would be great if it would be possible to import and export a .csv list. Just like the `redirect:import` and `redirect:export` command from the CLI

### LINKED ISSUE #8: Refactor fluid to Fusion templating
This can be similarly done like in the history controller of Neos and would lead to the following benefits:

* Easier to modify and extend
* We don't use form helpers for now but later replace the used form html tags with fusion forms in Neos 5.0 (planned)
* We want to have Fusion based backend modules in the long term

### LINKED ISSUE #12: Use dropdown to select redirect status code
Instead of having the number input that allows setting any kind of status code, it would be easier for editors and more robust to have a dropdown that allows to select status codes that make sense in a projects context.

Therefore it would be good to be able to configure a list of valid status codes including their (localizable) labels.

This list would then be the source for a dropdown that will be display instead of the number input.

By default, the status codes that are most commonly used should be preconfigured.

See also #6 for this.

### LINKED ISSUE #13: Allow setting a start & end date time for redirects in the UI
See https://github.com/neos/redirecthandler/issues/20

### PR DESCRIPTION
**Pull request by Sebobo**
This change does the following:

* Refactor view to fusion + react -> Resolves #8 
* Introduce new fields from RedirectHandler package into the UI -> Resolves #13 & #14 
* Introduces dropdown with configurable status codes -> Resolves #12 
* Adds the necessary views and buttons for importing and exporting redirects -> Resolves #3

This change depends on:

* https://github.com/neos/redirecthandler/pull/25
* https://github.com/neos/redirecthandler-neosadapter/pull/39
* https://github.com/neos/redirecthandler-databasestorage/pull/16

### CODE REVIEW

**Comment by Sebobo** [comment]
To review all packages the following composer config is needed:

```
"neos/redirecthandler": "dev-feature/introduce-new-fields as 3.0.0",
"neos/redirecthandler-databasestorage": "dev-feature/introduce-new-fields as 4.0.0",
"neos/redirecthandler-neosadapter": "dev-feature/introduce-new-fields as 4.0.0",
"neos/redirecthandler-ui": "dev-feature/introduce-new-fields as 2.0.0",
```

File: Resources/Private/FusionModule/Components/ImportProtocol.fusion [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        <table class={props.className}>
+            <tr>
+                <th>Status</th>
+                <th>Redirect</th>
+                <th>{I18n.translate('Neos.RedirectHandler.Ui:Modules:startDateTime')}</th>
+                <th>{I18n.translate('Neos.RedirectHandler.Ui:Modules:endDateTime')}</th>
+                <th>{I18n.translate('Neos.RedirectHandler.Ui:Modules:comment')}</th>
+            </tr>
+            <Neos.Fusion:Loop items={props.protocol} itemName="entry">
+                <tr class={props.className + '-entry--' + entry.type}>
```
Comment (hafofu) [inline Resources/Private/FusionModule/Components/ImportProtocol.fusion]: Is the `--` correct? Some lines underneath you use 1 `-`

File: Resources/Private/JavaScript/components/RedirectForm.tsx [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+                originalSourceUriPath: redirect ? redirect.sourceUriPath : null,
+                ...this.state,
+                startDateTime: startDateTime ? formatW3CString(new Date(startDateTime)) : null,
+                endDateTime: endDateTime ? formatW3CString(new Date(endDateTime)) : null,
+            }
+        };
+
+        this.setState({isSendingData: true});
+
+        fetch(redirect ? actions.update : actions.create, {
```
Comment (hafofu) [inline Resources/Private/JavaScript/components/RedirectForm.tsx]: I would move that logic out to a seperate "postRedirect" method.
E.g:
```
function postRedirect(path, body) {
  return fetch(path, {
    method: 'POST',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json; charset=UTF-8'
    },
    body: body && JSON.stringify(body)
  })
  .then(res => res.json())
  .then(async data => {
    if (data.success) { return data; }
    throw new Error(data.message);
  });
};
```

File: Resources/Private/JavaScript/components/RedirectForm.tsx [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+    /**
+     * Renders list of changed redirects to be used in a flash message
+     *
+     * @param changedRedirects
+     */
+    private renderChangedRedirects = (changedRedirects: Array<Redirect>): string => {
+        const {translate} = this.props;
+        return (
+            '<p>' + translate('message.relatedChanges', 'Related changes') + '</p>' +
```
Comment (hafofu) [inline Resources/Private/JavaScript/components/RedirectForm.tsx]: Never saw writing jsx as string representation.
Why not like that?
```
return (
  <> // Also called React.Fragment
    <p>{translate('message.relatedChanges', 'Related changes')}</p>
    <ul></ul>
  </>
);
```

File: Resources/Private/JavaScript/components/RedirectForm.tsx [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                               title={validSourceUriPathPattern} onChange={this.handleInputChange}
+                               autoFocus={true} required={true} placeholder="the-old-url/product-a"
+                               pattern={validSourceUriPathPattern} value={sourceUriPath || ''}/>
+                    </div>
+                    <div className="neos-control-group">
+                        <label className="neos-control-label"
+                               htmlFor={idPrefix + 'statusCode'}>{translate('statusCode')}</label>
+                        <select name="statusCode" id={idPrefix + 'statusCode'} value={statusCode}
+                                onChange={this.handleInputChange}>
+                            {Object.keys(statusCodes).map((code) => (
```
Comment (hafofu) [inline Resources/Private/JavaScript/components/RedirectForm.tsx]: You prefer
`(code) => (`
or
`code => (`
?

File: Resources/Private/JavaScript/index.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                ok: Function;
+                error: Function;
+                warning: Function;
+                info: Function;
+            },
+        };
+    }
+}
+
+(() => {
```
Comment (hafofu) [inline Resources/Private/JavaScript/index.tsx]: Is there a reason why you wrap that in a iif? (Immediately Invoked Function)
Your variables wont get added to window.

Alternative for a seperate scope is:
```
{
  const someVar = 'foobar';
}
```

File: webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                test: /\.(scss|css)$/,
+                use: [
+                    MiniCssExtractPlugin.loader,
+                    'css-loader?sourceMap',
+                    'sass-loader?sourceMap',
+                ],
+            },
+            {
+                enforce: 'pre',
+                test: /\.js$/,
```
Comment (hafofu) [inline webpack.config.js]: You dont resolve jsx, but added it to resolve.extensions

File: Resources/Private/FusionModule/Components/Redirect.List.fusion [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        }
+        create = Neos.Fusion:UriBuilder {
+            action = 'create'
+            format = 'json'
+        }
+    }
+
+    renderer = afx`
+        <div id="redirects-list-app"
+             data-show-hit-count={props.showHitCount ? 'true' : 'false'}
```
[the remark this answers is not in the document]
Comment (Sebobo) [reply Resources/Private/FusionModule/Components/Redirect.List.fusion]: No AFX will not render the attribute fully when it's value is `false`, so I don't get the expected type afterwards.

File: Resources/Private/JavaScript/components/RedirectForm.tsx (hunk 3 above)
Comment (Sebobo) [reply Resources/Private/JavaScript/components/RedirectForm.tsx]: I have to give a string to the FlashMessage api, therefore I cannot just give a ReactElement

File: Resources/Private/JavaScript/components/RedirectForm.tsx (hunk 4 above)
Comment (Sebobo) [reply Resources/Private/JavaScript/components/RedirectForm.tsx]: My IDE always adds the braces and sometimes I simply forget removing them ;)
So I prefer without. Thanks

File: Resources/Private/JavaScript/components/RedirectForm.tsx [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    sourceUriPath: '',
+    targetUriPath: '',
+    statusCode: -1,
+    startDateTime: '',
+    endDateTime: '',
+    comment: '',
+    isSendingData: false,
+};
+
+export class RedirectForm extends PureComponent<RedirectFormProps, RedirectFormState> {
```
[the remark this answers is not in the document]
Comment (Sebobo) [reply Resources/Private/JavaScript/components/RedirectForm.tsx]: Cool, yes let's try to refactor that tomorrow. Haven't used the new way yet.

File: Resources/Private/FusionModule/Components/ImportProtocol.fusion (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sebobo) [reply Resources/Private/FusionModule/Components/ImportProtocol.fusion]: Yes the class below was actually incorrect, thanks

File: Resources/Private/JavaScript/components/RedirectForm.tsx [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+
+        const parsedSourceUrl: URL = parseURL(sourceUriPath, host || location.origin);
+        const parsedTargetUrl: URL = parseURL(targetUriPath, location.origin);
+        if (parsedSourceUrl.pathname === parsedTargetUrl.pathname) {
+            notificationHelper.warning(translate('error.sameSourceAndTarget', 'The source and target paths cannot be the same'));
+            return;
+        }
+
+        const data = {
+            '__csrfToken': csrfToken,
```
[the remark this answers is not in the document]
Comment (Sebobo) [reply Resources/Private/JavaScript/components/RedirectForm.tsx]: nope

**Comment by Sebobo** [comment]
@jonnitto what kind of response do you get in the network tab in your browser when you get the error?

**Comment by jonnitto** [comment]
The Datepicker locks fine now…

**Comment by Sebobo** [comment]
@jonnitto you mean the ones for the status code explanations? 
They are a bit harder to translate and I didn't get to it yet. But we can add them later.

**Comment by jonnitto** [comment]
Of course. That‘s why I approve it.

**Comment by akappler** [comment]
Hi,
I have one case which can't be covered, yet. This might be something for the future so. 

We use the flowpack (https://github.com/Flowpack/neos-dimensionresolver) to have a top level domain as an dimension value (e.g. German_Swiss would usually be domain.com/de_ch/ but in our case domain.ch/de/ ). Therefore it would be awesome if you could set the target domain as well.

**Comment by Sebobo** [comment]
@akappler you can enter full urls in the target field, is that not enough?
By default the redirect will also just use the domain from the request which should cover most cases I think.
Or do you need a domain for each redirect and want to just make it easier to enter via the dropdown for the domain?

**Pull request merged**